File: examples/H20/robots/sdk_package/example/h2x/low_level/low_cmd.cpp
// low_cmd.cpp

#include <iostream>
#include <algorithm>
#include <thread>
#include <cmath>
#include <chrono>

#include "common/common_def.h"
#include "common/h2x_def.h"
#include "h2x/h2x_low_level.h"

using namespace humanoid;
using namespace humanoid::h2x;

constexpr double MAX_POS = M_PI * 10.0 / 180.0;
constexpr double DURATION = 10;      // unit: ms
constexpr double TOTAL_TIME = 2000;  // unit: ms

int main() {
  std::cout << "=== Low Level Command Test ===" << std::endl;
  H2xLowLevel low_level;
  LowCmd low_cmd;
  // start control loop
  constexpr size_t control_count = TOTAL_TIME / DURATION;
  for (size_t i = 0; i < control_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();
    auto pos = std::sin(M_PI * i * DURATION / 1000.0) * MAX_POS;
    auto vel = std::cos(M_PI * i * DURATION / 1000.0) * MAX_POS * M_PI;
    low_cmd.motor_cmd[0].mode = 1;
    low_cmd.motor_cmd[0].kp = 15.0;
    low_cmd.motor_cmd[0].kd = 0.5;
    low_cmd.motor_cmd[0].q = pos;
    low_cmd.motor_cmd[0].dq = vel;
    if (low_level.writeLowCmd(low_cmd) != 0) {
      std::cout << "pub low cmd failed!" << std::endl;
      return 0;
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = end - start;
    std::this_thread::sleep_for(
        std::chrono::microseconds(static_cast<int>(DURATION * 1000)) - elapsed);
  }

  return 0;
}

File: examples/H20/robots/sdk_package/example_py/h2x/low_level/low_cmd.py
#!/usr/bin/env python3

import humanoid_sdk_py
from humanoid_sdk_py import h2x
import time
import math


def main():
    print("=== Low Level Command Test ===")

    # Initialize objects
    low_level = h2x.H2xLowLevel()

    # Set low level command for joint 0
    MAX_POS = math.pi * 10.0 / 180.0
    DURATION = 10  # ms
    TOTAL_TIME = 2000  # ms
    control_count = TOTAL_TIME // DURATION
    low_cmd = h2x.LowCmd()

    for i in range(control_count):
        start_time = time.time()

        pos = math.sin(math.pi * i * DURATION / 1000.0) * MAX_POS
        vel = math.cos(math.pi * i * DURATION / 1000.0) * MAX_POS * math.pi
        
        # Set command for joint 0 (first joint)
        low_cmd.motor_cmd[0].mode = 1
        low_cmd.motor_cmd[0].kp = 15.0
        low_cmd.motor_cmd[0].kd = 0.5
        low_cmd.motor_cmd[0].q = pos
        low_cmd.motor_cmd[0].dq = vel

        if low_level.writeLowCmd(low_cmd) != 0:
            print("pub low cmd failed!")
            return 0
        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = DURATION - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    print("Low level command test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())
